akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/EventSourcedBehavior.scala

```diff
@@ -140,16 +139,28 @@ abstract class EventSourcedBehavior[Command, Event, State] private[akka] (
    * When persisting multiple events at once the snapshot is triggered after all the events have
    * been persisted.
    *
-   * Snapshots triggered by `snapshotWhen` will not trigger deletes of old snapshots and events if
+   * Snapshots triggered by `shouldSnapshot` will not trigger deletes of old snapshots and events if
    * [[EventSourcedBehavior.retentionCriteria]] with [[RetentionCriteria.snapshotEvery]] is used together with
    * `shouldSnapshot`. Such deletes are only triggered by snapshots matching the `numberOfEvents` in the
    * [[RetentionCriteria]].
    *
+   * Events can be deleted if `deleteEventsOnSnapshot` returns `true`.
+   *
    * @return `true` if snapshot should be saved at the given `state`, `event` and `sequenceNr` when the event has
    *         been successfully persisted
    */
   def shouldSnapshot(@unused state: State, @unused event: Event, @unused sequenceNr: Long): Boolean = false
 
+  /**
+   * Can be used to delete events after `shouldSnapshot`.
+   *
+   * Can be used in combination with `[[EventSourcedBehavior.retentionCriteria]]` in a way that events are triggered
+   * up the the oldest snapshot based on `[[RetentionCriteria.snapshotEvery]]` config.
+   *
+   * @return `true` if events should be deleted after `shouldSnapshot` evaluates to true
+   */
+  def deleteEventsOnSnapshot: Boolean = false
```

Comment:
For the Java API this is a bit weirder because it doesn't have context except for the API-docs. I think a user could easily end up missing the `shouldSnapshot` and just override just this expecting that it will delete events on snapshot.

```
public RetentionCriteria retentionCriteria() {
  return RetentionCriteria.snapshotEvery(100, 2);
}
public bool deleteEventsOnSnapshot() {
  return true;
}
```
I think that we need to either make either delete flag being true work or try to make this name more clear about being only for predicate based snapshot

Reply:
Initially the name of this method was `deleteEventsOnShouldSnapshot` which got changed into the current name.

I think because of the current implementation we should separate the two flags and their respective impact. 

The main problem here is the api of `def shouldSnapshot(@unused state: State, @unused event: Event, @unused sequenceNr: Long): Boolean` which should be evolved to a proper type, similar to `withRetention`. Maybe this is the way to go by adding `@deprecated` to the api and introducing a new method+type for this? This should then also be made for the scala api.

Reply:
Hmm, tricky to get it right without overdoing it. A type to return also means users has to figure out how to construct that, but maybe that is worth it because it makes what the flag refers to obvious.

Reply:
Maybe naming it back to `deleteEventsOnShouldSnapshot` @patriknw? ;-)

Reply:
Taking a step back, it's kind of weird to have the snapshot predicate and snapshot via RetentionCriteria as two separate things. We could add the snapshotWhen as another RetentionCriteria, and use the same approach for withDeleteEventsOnSnapshot.

Maybe that would loose the possibility to use both snapshotEvery(numberOfEvents) and snapshotWhen(predicate), but that is probably an unusual combination anyway.

Reply:
I like the idea of evolving the predicate api. But is a `snapshotWhen` predicate actually something that belongs to a `RetentionCriteria`? To me retention refers to the fact if a snapshot should be stored, and if so how many of them. I think we need a more composable api: 
- define when a snapshot is triggered (predicate and/or numberOfEvents)
- if triggered, what is the retention of snapshots?
- delete events after a snapshot is triggered

To me this are three different properties that may be composed as needed (with some validation). But this probably goes into "not overdoing it".

For now I will change this PR and put `snapshotWhen` into `RetentionCriteria` to move it forward, we can see?

Reply:
I think it's fine that we add some more for when to snapshot in RetentionCriteria. That is anyway the normal way to define that snapshots should be enabled.

Originally I think retention referred to deletion of events, but we didn't allow deletion unless there are snapshots. Yeah, maybe not perfect name, but that is what it is.

> For now I will change this PR and put snapshotWhen into RetentionCriteria to move it forward

Agreed

Reply:
Do you agree that the new `snapshotWhen` should also receive a `keepNSnapshots` param, like `def snapshotEvery(numberOfEvents: Int, keepNSnapshots: Int): SnapshotCountRetentionCriteria`? 

If not, after this pr, it will not be possible anymore to delete snapshots when using the (new) predicate based api because we remove option to combine snapshotEvery and predicate ways to define snapshot triggering.

Reply:
Or we rely on `only-one-snapshot` to clean up snapshots?